src/gfx/convert.cpp

```diff
@@ -0,0 +1,831 @@
+			++i;
+		}
+		options.verbosePrint("]\n");
+	}
+
+	// Now, iterate through the tiles, generating proto-palettes as we go
+	// We do this unconditionally because this performs the image validation (which we want to
+	// perform even if no output is requested), and because it's necessary to generate any
+	// output (with the exception of an un-duplicated tilemap, but that's an acceptable loss.)
+	std::vector<ProtoPalette> protoPalettes;
+	DefaultInitVec<AttrmapEntry> attrmap{};
+
+	for (auto tile : png.visitAsTiles(options.columnMajor)) {
+		ProtoPalette tileColors;
+		AttrmapEntry &attrs = attrmap.emplace_back();
+
+		for (uint32_t y = 0; y < 8; ++y) {
+			for (uint32_t x = 0; x < 8; ++x) {
+				tileColors.add(tile.pixel(x, y));
+			}
+		}
+
+		// Insert the palette, making sure to avoid overlaps
+		// TODO: if inserting (0, 1), (0, 2), and then (0, 1, 2), we might have a problem!
+		for (size_t i = 0; i < protoPalettes.size(); ++i) {
+			switch (tileColors.compare(protoPalettes[i])) {
+			case ProtoPalette::WE_BIGGER:
+				protoPalettes[i] = tileColors; // Override them
+				[[fallthrough]];
+			case ProtoPalette::THEY_BIGGER:
+				// Do nothing, they already contain us
+				attrs.protoPaletteID = i;
+				goto contained;
+			case ProtoPalette::NEITHER:
+				break; // Keep going
+			}
+		}
+		attrs.protoPaletteID = protoPalettes.size();
+		protoPalettes.push_back(tileColors);
+contained:;
+	}
+
+	options.verbosePrint("Image contains %zu proto-palette%s\n", protoPalettes.size(),
+	                     protoPalettes.size() != 1 ? "s" : "");
+
+	// Sort the proto-palettes by size, which improves the packing algorithm's efficiency
+	// TODO: try keeping the palettes stored while inserting them instead, might perform better
+	std::sort(
+		protoPalettes.begin(), protoPalettes.end(),
+		[](ProtoPalette const &lhs, ProtoPalette const &rhs) { return lhs.size() < rhs.size(); });
+
+	// Run a "pagination" problem solver
+	// TODO: allow picking one of several solvers?
+	auto [mappings, nbPalettes] = packing::overloadAndRemove(protoPalettes);
+	assert(mappings.size() == protoPalettes.size());
+	if (options.beVerbose) {
+		options.verbosePrint("Proto-palette mappings: (%zu palettes)\n", nbPalettes);
+		for (size_t i = 0; i < mappings.size(); ++i) {
+			options.verbosePrint("%zu -> %zu\n", i, mappings[i]);
+		}
+	}
+
+	// TODO: optionally, "decant" the result
+
+	// Generate the actual palettes from the mappings
+	std::vector<Palette> palettes(nbPalettes);
+	for (size_t protoPalID = 0; protoPalID < mappings.size(); ++protoPalID) {
+		auto &pal = palettes[mappings[protoPalID]];
+		for (uint16_t color : protoPalettes[protoPalID]) {
+			pal.addColor(color);
+		}
+	}
+
+	// "Sort" colors in the generated palettes, see the man page for the flowchart
+	auto [palSize, palRGB, palAlpha] = png.getEmbeddedPal();
+	if (palRGB) {
+		sorting::indexed(palettes, palSize, palRGB, palAlpha);
+	} else if (png.hasNonGray()) {
```

Comment:
This is the only call to `hasNonGray` I see (`bool hasNonGray() const { return !isGrayOnly; }`); can we avoid the extra negative and do `bool isGrayscale() const { return isGrayOnly; }`?

Reply:
I wanted to avoid the term "grayscale" because it might be confused with "grayscale PNGs", which are not the point of this check

Reply:
`isGrayOnly()`, then, and make the boolean member field be `_isGrayOnly`?

Reply:
I am changing the function anyway:
```cxx
	bool isSuitableGrayscale() const {
		if (!isGrayOnly) {
			return false;
		}
		// Check that all of the grays don't fall into the same "bin"
		if (colors.size() > options.maxPalSize()) { // Apply the Pigeonhole Principle
			return false;
		}
		uint8_t bins = 0;
		for (auto const & color : colors) {
			if (color.isTransparent()) {
				continue;
			}
			uint8_t mask = 1 << color.grayIndex();
			if (bins & mask) { // Two in the same bin!
				return false;
			}
			bins |= mask;
		}
		return true;
	}
```

Reply:
Grammatically "`isSuitableForGrayscale`", but sure.

Reply:
I was intending for `isASuitableGrayscale`, but I guess your way fits the context better ("is suitable for grayscale [sorting method]")